ast/property.go

```diff
@@ -78,10 +78,192 @@ func (p *PropertySubscript) rootName() string {
 	return p.Index.(string)
 }
 
-func terminatesName(c byte) bool {
+type propertyAccessParser struct {
+	parent    syntax.Node
+	text      string
+	accessors []PropertyAccessor
+	diags     syntax.Diagnostics
+}
+
+func (p *propertyAccessParser) error(msg string) {
+	p.diags.Extend(syntax.NodeError(p.parent, msg))
+}
+
+func (p *propertyAccessParser) errorf(f string, args ...any) {
+	p.error(fmt.Sprintf(f, args...))
+}
+
+func (p *propertyAccessParser) terminatesName(c byte) bool {
 	return c == '.' || c == '[' || c == '}' || unicode.IsSpace(rune(c))
 }
 
+// Appends a property accessor.
+func (p *propertyAccessParser) append(accessor PropertyAccessor) {
+	p.accessors = append(p.accessors, accessor)
+}
+
+// Consumes a byte of input. Use peek() prior to next() to determine what the byte is if one is
+// available.
+func (p *propertyAccessParser) next() {
+	p.text = p.text[1:]
+}
+
+// Returns (but does not consume) the next byte of input, if any.
+func (p *propertyAccessParser) peek() (byte, bool) {
+	if len(p.text) == 0 {
+		return 0, false
+	}
+	return p.text[0], true
+}
+
+// Parses a property access. See the comment on parsePropertyAccess for the grammar and examples.
+func (p *propertyAccessParser) parse() (string, *PropertyAccess, syntax.Diagnostics) {
+	for {
+		c, ok := p.peek()
+		if !ok {
+			p.error("unterminated interpolation")
+			return p.finish()
+		}
+
+		switch c {
+		case '}':
+			p.next()
+			return p.finish()
+		case '.':
+			if len(p.accessors) == 0 {
+				p.error("the root property must be a string subscript or a name")
+			}
+			p.next()
+			p.append(p.parseName())
+		case '[':
+			p.next()
+			p.append(p.parseSubscript())
+		default:
+			if unicode.IsSpace(rune(c)) {
+				p.error("unterminated interpolation")
+				return p.finish()
+			}
+			p.append(p.parseName())
+		}
+	}
+}
+
+// Terminates parsing. If there are no accessors (e.g. `${` or `${}`), appends an empty property name
+// accessor. Returns the rest of the string, the access, and any diagnostics.
+func (p *propertyAccessParser) finish() (string, *PropertyAccess, syntax.Diagnostics) {
+	if len(p.accessors) == 0 {
+		p.append(&PropertyName{Name: ""})
+	}
+
+	rest := p.text
+	access := &PropertyAccess{Accessors: p.accessors}
+	return rest, access, p.diags
+}
+
+// Parses a property name (e.g. `foo`).
+func (p *propertyAccessParser) parseName() *PropertyName {
+	var b strings.Builder
+	for {
+		c, ok := p.peek()
+		if !ok || p.terminatesName(c) {
+			break
+		}
+		p.next()
+		b.WriteByte(c)
+	}
+	if b.Len() == 0 {
+		p.errorf("missing property name")
+	}
+	return &PropertyName{Name: b.String()}
+}
+
+// Parses a subscript accessor (e.g. `["foo"]` or `[1]`).
+//
+// At this point we are already past the opening `[`. Consumes the terminating `]`, if any.
+func (p *propertyAccessParser) parseSubscript() *PropertySubscript {
+	c, ok := p.peek()
+	if !ok {
+		p.error("missing closing bracket in subscript")
```

Comment:
```suggestion
		p.error("missing closing bracket ']' in subscript")
```

What do you think about including the character in the error message to avoid any ambiguity? Same applies to the other error messages that indicate a specific missing character/sequence.